File: zilli/privacy/consent.py
from __future__ import annotations

import logging
import time
from dataclasses import dataclass
from enum import Enum
from typing import Optional

logger = logging.getLogger("zilli.privacy.consent")


class ConsentStatus(str, Enum):
    GRANTED = "granted"
    DENIED = "denied"
    EXPIRED = "expired"
    REVOKED = "revoked"
    PENDING = "pending"


class DataUse(str, Enum):
    LOCAL_INFERENCE = "local_inference"
    CLOUD_INFERENCE = "cloud_inference"
    TRAINING = "training"
    DISTILLATION = "distillation"
    AUDIT = "audit"
    ANONYMIZED_ANALYTICS = "anonymized_analytics"


@dataclass
class ConsentRecord:
    tenant_id: str
    user_id: str
    data_use: DataUse
    status: ConsentStatus
    granted_at: float
    expires_at: float
    revoked_at: Optional[float] = None
    purpose: str = ""


class ConsentManager:
    def __init__(self):
        self._records: list[ConsentRecord] = []

    def grant(self, tenant_id: str, user_id: str, data_use: DataUse,
              ttl_days: int = 365, purpose: str = "") -> ConsentRecord:
        now = time.time()
        record = ConsentRecord(
            tenant_id=tenant_id,
            user_id=user_id,
            data_use=data_use,
            status=ConsentStatus.GRANTED,
            granted_at=now,
            expires_at=now + ttl_days * 86400,
            purpose=purpose,
        )
        self._records.append(record)
        logger.info("Consent granted: %s/%s for %s", tenant_id, user_id, data_use.value)
        return record

    def revoke(self, tenant_id: str, user_id: str, data_use: DataUse) -> bool:
        for r in reversed(self._records):
            if (r.tenant_id == tenant_id and r.user_id == user_id
                    and r.data_use == data_use and r.status == ConsentStatus.GRANTED):
                r.status = ConsentStatus.REVOKED
                r.revoked_at = time.time()
                logger.info("Consent revoked: %s/%s for %s", tenant_id, user_id, data_use.value)
                return True
        return False

    def check(self, tenant_id: str, user_id: str, data_use: DataUse) -> bool:
        now = time.time()
        for r in reversed(self._records):
            if (r.tenant_id == tenant_id and r.user_id == user_id
                    and r.data_use == data_use):
                if r.status == ConsentStatus.GRANTED and r.expires_at > now:
                    return True
                if r.status == ConsentStatus.REVOKED:
                    return False
        return False

    def list_for_user(self, tenant_id: str, user_id: str) -> list[ConsentRecord]:
        return [
            r for r in self._records
            if r.tenant_id == tenant_id and r.user_id == user_id
        ]

    def list_active(self, tenant_id: str) -> list[ConsentRecord]:
        now = time.time()
        return [
            r for r in self._records
            if r.tenant_id == tenant_id
            and r.status == ConsentStatus.GRANTED
            and r.expires_at > now
        ]

    def expire_all(self, tenant_id: str, user_id: str):
        now = time.time()
        for r in self._records:
            if r.tenant_id == tenant_id and r.user_id == user_id:
                if r.status == ConsentStatus.GRANTED and r.expires_at > now:
                    r.expires_at = now
                    r.status = ConsentStatus.EXPIRED
                    logger.info("Consent expired: %s/%s for %s", tenant_id, user_id, r.data_use.value)

File: zilli-rs/src/privacy/consent.rs
use std::sync::Arc;
use chrono::{DateTime, Utc};
use parking_lot::RwLock;
use serde::{Deserialize, Serialize};
use strum::{Display, EnumString};

#[derive(Debug, Clone, Serialize, Deserialize, Display, EnumString, PartialEq)]
pub enum ConsentStatus {
    Granted,
    Denied,
    Expired,
    Revoked,
    Pending,
}

#[derive(Debug, Clone, Serialize, Deserialize, Display, EnumString, PartialEq, Hash, Eq)]
pub enum DataUse {
    LocalInference,
    CloudInference,
    Training,
    Distillation,
    Audit,
    AnonymizedAnalytics,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConsentRecord {
    pub user_id: String,
    pub data_use: DataUse,
    pub status: ConsentStatus,
    pub granted_at: DateTime<Utc>,
    pub expires_at: Option<DateTime<Utc>>,
    pub purpose: Option<String>,
    pub revoked_at: Option<DateTime<Utc>>,
}

pub struct ConsentManager {
    records: Arc<RwLock<Vec<ConsentRecord>>>,
}

impl ConsentManager {
    pub fn new() -> Self {
        Self {
            records: Arc::new(RwLock::new(Vec::new())),
        }
    }
}

impl Clone for ConsentManager {
    fn clone(&self) -> Self {
        Self {
            records: Arc::clone(&self.records),
        }
    }
}

impl ConsentManager {

    pub fn grant(
        &self,
        user_id: &str,
        data_use: DataUse,
        purpose: Option<&str>,
        ttl_days: Option<i64>,
    ) -> ConsentRecord {
        let record = ConsentRecord {
            user_id: user_id.to_string(),
            data_use,
            status: ConsentStatus::Granted,
            granted_at: Utc::now(),
            expires_at: ttl_days.map(|d| Utc::now() + chrono::Duration::days(d)),
            purpose: purpose.map(|s| s.to_string()),
            revoked_at: None,
        };
        self.records.write().push(record.clone());
        record
    }

    pub fn revoke(&self, user_id: &str, data_use: &DataUse) -> bool {
        let mut records = self.records.write();
        for record in records.iter_mut().rev() {
            if record.user_id == user_id && record.data_use == *data_use {
                record.status = ConsentStatus::Revoked;
                record.revoked_at = Some(Utc::now());
                return true;
            }
        }
        false
    }

    pub fn check(&self, user_id: &str, data_use: &DataUse) -> ConsentStatus {
        let records = self.records.read();
        for record in records.iter().rev() {
            if record.user_id == user_id && record.data_use == *data_use {
                match &record.status {
                    ConsentStatus::Granted => {
                        if let Some(expires) = record.expires_at {
                            if Utc::now() > expires {
                                return ConsentStatus::Expired;
                            }
                        }
                        return ConsentStatus::Granted;
                    }
                    other => return other.clone(),
                }
            }
        }
        ConsentStatus::Pending
    }

    pub fn get_user_records(&self, user_id: &str) -> Vec<ConsentRecord> {
        self.records
            .read()
            .iter()
            .filter(|r| r.user_id == user_id)
            .cloned()
            .collect()
    }

    pub fn get_active_consents(&self, data_use: &DataUse) -> Vec<ConsentRecord> {
        let now = Utc::now();
        self.records
            .read()
            .iter()
            .filter(|r| {
                r.data_use == *data_use
                    && r.status == ConsentStatus::Granted
                    && r.expires_at.map_or(true, |e| e > now)
            })
            .cloned()
            .collect()
    }
}

impl Default for ConsentManager {
    fn default() -> Self {
        Self::new()
    }
}
